include/deal.II/lac/vector_space_vector.h

```diff
@@ -0,0 +1,219 @@
+
+    /**
+     * Assignement <tt>*this = a*V</tt>.
+     */
+    virtual void equ(const Number a, const VectorSpaceVector<Number> &V) = 0;
+
+    /**
+     * Returns the l<sub>1</sub> norm of the vector (i.e., the sum of the
+     * absolute values of all entries among all processors).
+     */
+    virtual real_type l1_norm() = 0;
+
+    /**
+     * Returns the l<sub>2</sub> norm of the vector (i.e., the square root of
+     * the sum of the square of all entries among all processors).
+     */
+    virtual real_type l2_norm() = 0;
+
+    /**
+     * Returns the maximum norm of the vector (i.e., the maximum absolute value
+     * among all entries and among all processors).
+     */
+    virtual real_type linfty_norm() = 0;
+
+    /**
+     * Performs a combined operation of a vector addition and a subsequent
+     * inner product, returning the value of the inner product. In other
+     * words, the result of this function is the same as if the user called
+     * @code
+     * this->add(a, V);
+     * return_value = *this * W;
+     * @endcode
+     *
+     * The reason this function exists is that this operation involves less
+     * memory transfer than calling the two functions separately. This
+     * method only needs to load three vectors, @p this, @p V, @p W, whereas
+     * calling separate methods means to load the calling vector @p this
+     * twice. Since most vector operations are memory transfer limited, this
+     * reduces the time by 25\% (or 50\% if @p W equals @p this).
+     */
+    virtual Number add_and_dot(const Number a,
+                               const VectorSpaceVector<Number> &V,
+                               const VectorSpaceVector<Number> &W) = 0;
+
+    /**
+     * Returns the global size of the vector, equal to the sum of the number of
+     * locally owned indices amon all processors.
+     */
+    virtual size_type size() const = 0;
+
+    /**
+     * Return an index set that describes which elements of this vector are
+     * owned by the current processor. As a consequence, the index sets returned
+     * on different procesors if this is a distributed vector will form disjoint
+     * sets that add up to the complete index set. Obviously, if a vector is
+     * created on only one processor, then the result would satisfy
+     * @code
+     *  vec.locally_owned_elements() == complete_index_set(vec.size())
+     * @endcode
+     */
+    virtual IndexSet locally_owned_elements() const = 0;
+
+    /**
+     * Swap the contents of this vector and the other vector @p V. One could
+     * do this operation with a temporary variable and copying over the data
+     * elements, but this function is significantly more efficient since it
+     * only swaps the pointers to the data of the two vectors and therefore
+     * does not need to allocate temporary storage and move data around.
+     */
+    virtual void swap(VectorSpaceVector<Number> &V) = 0;
```

Comment:
I don't think you need this -- swapping objects only makes sense if you actually know the concrete data types of both objects.